src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1138 @@
+{
+   for(auto& effect : mRealtimeGroupProcessors)
+      effect->mAudioProcessor->setProcessing(false);
+   return true;
+}
+
+bool VST3Effect::RealtimeResume()
+{
+   for(auto& effect : mRealtimeGroupProcessors)
+      effect->mAudioProcessor->setProcessing(true);
+   return true;
+}
+
+bool VST3Effect::RealtimeProcessStart()
+{
+   wxASSERT(mPendingChanges == nullptr);
+
+   if(mComponentHandler != nullptr)
+      //Same parameter changes are used among all of the relatime processors
+      mPendingChanges = mComponentHandler->getPendingChanges();
+   return true;
+}
+
+size_t VST3Effect::RealtimeProcess(int group, float** inBuf, float** outBuf, size_t numSamples)
+{
+   auto& effect = mRealtimeGroupProcessors[group];
+   return VST3ProcessBlock(effect->mEffectComponent.get(), effect->mSetup, inBuf, outBuf, numSamples, mPendingChanges.get());
+}
+
+bool VST3Effect::RealtimeProcessEnd()
+{
+   mPendingChanges.reset();
+   return true;
+}
+
+int VST3Effect::ShowClientInterface(wxWindow& parent, wxDialog& dialog, bool forceModal)
+{
+   if(!IsGraphicalUI())
+   {
+      //Restrict resize of the "plain" dialog
+      dialog.SetMaxSize(dialog.GetSize());
+      dialog.SetMinSize(dialog.GetSize());
+   }
+   if(forceModal)
+      return dialog.ShowModal();
+
+   dialog.Show();
+   return 0;
+}
+
+bool VST3Effect::SetHost(EffectHostInterface* host)
+{
+   mEffectHost = host;
+   return true;
+}
+
+bool VST3Effect::IsGraphicalUI()
+{
+   return mPlugView != nullptr;
+}
+
+bool VST3Effect::PopulateUI(ShuttleGui& S)
+{
+   using namespace Steinberg;
+
+   mParent = S.GetParent();
+
+   if(mComponentHandler != nullptr)
+   {
+      SyncParameters();
+
+      auto parent = S.GetParent();
+      if(GetType() == EffectTypeGenerate)
+      {
+         auto vSizer = std::make_unique<wxBoxSizer>(wxVERTICAL);
+         auto controllsRoot = std::make_unique<wxWindow>(parent, wxID_ANY);
+         if(!LoadVSTUI(controllsRoot.get()))
+            VST3Utils::BuildPlainUI(controllsRoot.get(), mEditController, mComponentHandler);
+         vSizer->Add(controllsRoot.release());
+
+         auto duration = std::make_unique<NumericTextCtrl>(
+               parent, wxID_ANY,
+               NumericConverter::TIME,
+               mEffectHost->GetDurationFormat(),
+               mEffectHost->GetDuration(),
+               mSetup.sampleRate,
+               NumericTextCtrl::Options{}
+                  .AutoPos(true)
+            );
+
+         mDuration = duration.get();
+
+         auto hSizer = std::make_unique<wxBoxSizer>(wxHORIZONTAL);
+         hSizer->Add(safenew wxStaticText(parent, wxID_ANY, _("Duration:")));
+         hSizer->AddSpacer(5);
+         hSizer->Add(duration.release());
+         vSizer->AddSpacer(10);
+         vSizer->Add(hSizer.release());
+
+         parent->SetMinSize(vSizer->CalcMin());
+         parent->SetSizer(vSizer.release());
+      }
+      else if(!LoadVSTUI(parent))
+      {
+         VST3Utils::BuildPlainUI(
```

Comment:
There are two calls to `BuildPlainUI` here.  They pass pointers to mEditController and mComponentHandler.  A look inside the function shows that it is assumed that the pointers are not null, but there is no comment to that effect.  That condition is guaranteed here, so there is not a bug.

Still, I would prefer to say what I mean with types: make `BuildPlainUI` take references instead of pointers to make clear that null is not allowed.  Then at the places where we call it, we have to dereference pointers.  That would make me look for the proof that the pointers are not null, and I would get it from the suggested comments in the header about invariants.